src/discljord/messaging/specs.clj

```diff
@@ -205,6 +224,121 @@
                                             :message-reference/channel_id
                                             :message-reference/guild_id]))
 
+(s/def ::application-id ::ds/application-id)
+
+(def command-option-types
+  {:sub-command 1
+   :sub-command-group 2
+   :string 3
+   :integer 4
+   :boolean 5
+   :user 6
+   :channel 7
+   :role 8})
+
+(s/def :command.option/type (set (vals command-option-types)))
+
+(s/def :command.option/name (string-spec #"[\w-]{1,32}"))
+(s/def :command.option/description (string-spec 1 100))
+(s/def :command.option/default boolean?)
+(s/def :command.option/required boolean?)
+
+(s/def :command.option.choice/name (string-spec 1 100))
+
+(s/def :command.option.choice/value
+  (s/or :string string? :int int?))
+
+(s/def :command.option/choice (s/keys :req-un [:command.option.choice/name
+                                               :command.option.choice/value]))
+
+(s/def :command.option/choices (s/coll-of :command.option/choice))
+
+(s/def :command/option (s/and (s/keys :req-un [:command.option/type
+                                               :command.option/name
+                                               :command.option/description]
+                                      :opt-un [:command.option/default
+                                               :command.option/required
+                                               :command.option/choices
+                                               :command.option/options])
+                              #(<= (count (:choices %)) 25)
+                              #(not-any? #{(command-option-types :sub-command-group)} (map :type (:options %)))
+                              #(or (= (command-option-types :sub-command-group) (:type %))
+                                   (not-any? #{(command-option-types :sub-command)} (map :type (:options %))))))
+
+(def command-permission-types
+  {:role 1
+   :user 2})
+
+(s/def :command.permission/id ::ds/snowflake)
+
+(s/def :command.permission/type (set (vals command-permission-types)))
+
+(s/def :command.permission/permission boolean?)
+
+(s/def :command/permission (s/keys :req-un [:command.permission/id
+                                            :command.permission/type
+                                            :command.permission/permission]))
```

Comment:
To match with the other changes, this needs to have `discljord.messaging.specs.command.{option/permission/whatever}`, as unfortunate as it is that it's such a long spec. I recommend putting the spec on a different line than the spec's name.

Reply:
I've done it like this mainly for consistency. This goes for every spec in there right now:
- If it's a top level spec (or it's used at the top level), it has the `discljord.messaging.specs` namespace prefix
- If it's only used to compose other spec, it doesn't have the prefix

So if you want this to be changed, should I do it for every spec in the namespace (might be a breaking change, hence why I decided against it initially)?

Reply:
I was thinking about this as I looked at some of the other specs in other namespaces. I think that leaving it as-is is fine. I'm going to approve this PR. That said, I want to think about renaming all the internal specs to include bigger namespaces so that there is no conflict if the user wants to use short namespaces for their own specs that conflict.